### PR TITLE
Add code action for access modifier of 'main' function 

### LINKED ISSUE #30538: Add a quick fix for the main function
**Description:**
It would be good if we can have a quick fix for the 'the main function should be public' diagnostic.
Screenshot 2021-05-13 at 14 24 00

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gayalkuruppu**
quick fix changes this,

Screenshot from 2021-06-28 17-00-37

to this,

Screenshot from 2021-06-28 17-00-50

Fixes #30538 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (mohanvive) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java:2]: Change the license year.

Comment (mohanvive) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java:2]: Check other files as well.

**Comment by mohanvive** [comment]
@gayalkuruppu Change the PR title and commit message. message "quick fix" not helps.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Code Action for adding access modifiers.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class AccessModifierCodeAction extends AbstractCodeActionProvider {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java:42]: Should be `AddAccessModifierCodeAction`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static final String NAME = "Add Access Modifier";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic, 
+                                                    DiagBasedPositionDetails positionDetails, 
+                                                    CodeActionContext context) {
+        if (!(diagnostic.message().contains(CommandConstants.NO_PUBLIC_FOR_MAIN_FUNCTION))) {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java:50]: Shall we use the diagnostic code instead. `diagnostic.diagnosticInfo().code()`.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AccessModifierTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test Cases for CodeActions.
+ *
+ * @since 2.0.0
+ */
+public class AccessModifierTest extends AbstractCodeActionTest {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AccessModifierTest.java:29]: Should be `AddAccessModifierTest`.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AccessModifierTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public String getResourceDir() {
+        return "access-modifier";
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/AccessModifierTest.java:33]:
```suggestion
        return "add-access-modifier";
```
Resource directory should be renamed as well

File: language-server/modules/langserver-core/src/test/resources/codeaction/access-modifier/source/make_main_function_public_source2.bal [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class TestClass{
+    function main(){
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/resources/codeaction/access-modifier/source/make_main_function_public_source2.bal:2]:
```suggestion
    function main() {
```

File: language-server/modules/langserver-core/src/test/resources/codeaction/access-modifier/source/make_main_function_public_source2.bal [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class TestClass{
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/resources/codeaction/access-modifier/source/make_main_function_public_source2.bal:1]:
```suggestion
class TestClass {
```

**Review by IMS94 — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java (hunk 3 above)
Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AccessModifierCodeAction.java]:
```suggestion
        if (!(diagnostic.diagnosticInfo().code().equals("BCE2096"))){
```
changed to this

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class AddAccessModifierCodeAction extends AbstractCodeActionProvider {
+
+    public static final String NAME = "Add Access Modifier";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic, 
+                                                    DiagBasedPositionDetails positionDetails, 
+                                                    CodeActionContext context) {
+        if (!(diagnostic.diagnosticInfo().code().equals("BCE2096"))) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java:50]: shall we use `!("BCE2096".equals(diagnostic.diagnosticInfo().code()))`

**Review by malinthar — CHANGES_REQUESTED** [review]

**Review by malinthar — DISMISSED** [review]
LGTM

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java [hunk 9]
```diff
@@ -129,4 +129,6 @@
     public static final String OPTIMIZE_IMPORTS_TITLE = "Optimize all imports";
     
     public static final String REPORT_USAGE_STATISTICS_COMMAND_TITLE = "Report usage statistics";
+    
+    public static final String MAKE_MAIN_FUNCTION_PUBLIC = "Make 'main' function public";
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:133]: What if we use `Convert to public function`? @IMS94 @gayalkuruppu @mohanvive  WDYT?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic, 
+                                                    DiagBasedPositionDetails positionDetails, 
+                                                    CodeActionContext context) {
+        if (!(DiagnosticErrorCode.MAIN_SHOULD_BE_PUBLIC.diagnosticId().equals(diagnostic.diagnosticInfo().code()))) {
+            return Collections.emptyList();
+        }
+        
+        Optional<FunctionDefinitionNode> funcDef = CodeActionUtil.getEnclosedFunction(positionDetails.matchedNode());
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java:55]: Shall we do an `isEmpty` check as well for safety?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+        
+        String editText = "public ";
+        List<CodeAction> codeActions = new ArrayList<>();
+        List<TextEdit> edits = new ArrayList<>();
+        edits.add(new TextEdit(new Range(funcBodyStart, funcBodyStart), editText));
+
+        String commandTitle = CommandConstants.MAKE_MAIN_FUNCTION_PUBLIC;
+        codeActions.add(createQuickFixCodeAction(commandTitle, edits, context.fileUri()));
+        
+        return codeActions;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddAccessModifierCodeAction.java:66]: We don't need to explicitly create a list and add the code action here. Instead you can use `Arrays.asList` API. Same applies for line 61.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java (hunk 9 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:133]: +1

Comment (gayalkuruppu) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:133]: > What if we use `Convert to public function`? @IMS94 @gayalkuruppu @mohanvive WDYT?

```suggestion
    public static final String CONVERT_FUNC_PUBLIC = "Convert to public function";
```
or

```suggestion
    public static final String MAKE_MAIN_FUNCTION_PUBLIC = "Convert to public function";
```
Should I change the Variable name as well? BTW we only make the **main function public**

Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/constants/CommandConstants.java:133]: +1 to change the variable name as well

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]
LGTM!

**Pull request merged**